### Model Evaluation Exercises



2. Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

In [15]:
Accuracy = (46 + 34)/(46 + 7 + 13 + 34)
print(f'Accuracy = {Accuracy}')

Accuracy = 0.8


In [16]:
Recall = (46 / (46 + 13))
print(f'Recall = {Recall}')

Recall = 0.7796610169491526


In [17]:
Precision = (46/(46 + 7))
print(f'Precision = {Precision}')

Precision = 0.8679245283018868


- In the context of this problem, what is a false positive?

A false positive would be you predict cat but it is actually a dog.

- In the context of this problem, what is a false negative?

A false negative would be you predict a dog but it is actually a cat. 

- How would you describe this model?

In [37]:
Accuracy = (46 + 34)/(46 + 7 + 13 + 34)
print(f'Accuracy = {Accuracy}')

baseline_accuracy = (46 /100)
print(f'Baseline Accuracy = {baseline_accuracy}')

Accuracy = 0.8
Baseline Accuracy = 0.46


In [39]:
Recall = (46 / (46 + 7))
print(f'Recall = {Recall}')

baseline_recall = (46 / 46)
print(f'Baseline Recall = {baseline_recall}')

Recall = 0.8679245283018868
Baseline Recall = 1.0


In [41]:
Precision = (46/(46 + 13))
print(f'Precision = {Precision}')

baseline_precision = (46/100)
print(f'Baseline Precision = {baseline_precision}')

Precision = 0.7796610169491526
Baseline Precision = 0.46


- Overall it is a fairly accurate model. It's 80% accurate. It has slightly higher recall than precision. Meaning, it errors on the side of guessing dog instead of cat, so it doesn't miss out on any dog predictions, at the cost of having slightly less accurate dog predisctions. 

### 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

In [43]:
import pandas as pd

In [112]:
df = pd.read_csv('c3.csv')
df

actual     model1     model2     model3
0    No Defect  No Defect     Defect  No Defect
1    No Defect  No Defect     Defect     Defect
2    No Defect  No Defect     Defect  No Defect
3    No Defect     Defect     Defect     Defect
4    No Defect  No Defect     Defect  No Defect
..         ...        ...        ...        ...
195  No Defect  No Defect     Defect     Defect
196     Defect     Defect  No Defect  No Defect
197  No Defect  No Defect  No Defect  No Defect
198  No Defect  No Defect     Defect     Defect
199  No Defect  No Defect  No Defect     Defect

[200 rows x 4 columns]

In [113]:
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [64]:
df.describe()

actual     model1     model2     model3
count         200        200        200        200
unique          2          2          2          2
top     No Defect  No Defect  No Defect  No Defect
freq          184        190        110        101

An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. 

MODEL - 
* positive - defect (target is positive)
* negative - no defect 
BASELINE - most frequent category (no defect)




* TP - Defect identified, defect found
* TN - No defect identified, no defect present
* FP - Defect identified, no defect present
* FN - No defect identified, defect present

In [82]:
df['baseline'] = 'Defect'

In [83]:
df.head()

actual     model1  model2     model3 baseline
0  No Defect  No Defect  Defect  No Defect   Defect
1  No Defect  No Defect  Defect     Defect   Defect
2  No Defect  No Defect  Defect  No Defect   Defect
3  No Defect     Defect  Defect     Defect   Defect
4  No Defect  No Defect  Defect  No Defect   Defect

### Which evaluation metric would be appropriate here? 

- Recall would be appropriate because a false negative, of not identifying a defect that exsists has the greatest cost. 

In [87]:
# Model 1

positive = 'Defect'

model_accuracy = (df.model1 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model1  == positive]
model_precision = (subset.model1  == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model1  == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model 1 accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()
print(f'   model 1 recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')
print()
print(f'   model 1 precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

   model 1 accuracy: 95.00%
baseline accuracy: 8.00%

   model 1 recall: 50.00%
baseline recall: 100.00%

   model 1 precision: 80.00%
baseline precision: 8.00%


In [ ]:
### Finish later
'''


###. def model_evaluation(column):
    
    positive = 'Defect'

    model_accuracy = (df[column] == df.actual).mean()
    baseline_accuracy = (df.baseline == df.actual).mean()'''

In [85]:
# Model 2

positive = 'Defect'

model_accuracy = (df.model2 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model1  == positive]
model_precision = (subset.model2  == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model2  == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model 2 accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()
print(f'   model 2 recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')
print()
print(f'   model 2 precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

   model 2 accuracy: 56.00%
baseline accuracy: 8.00%

   model 2 recall: 56.25%
baseline recall: 100.00%

   model 2 precision: 40.00%
baseline precision: 8.00%


In [86]:
# model 3

positive = 'Defect'

model_accuracy = (df.model3 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model3  == positive]
model_precision = (subset.model3  == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model3  == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model 3 accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()
print(f'   model 3 recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')
print()
print(f'   model 3 precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

   model 3 accuracy: 55.50%
baseline accuracy: 8.00%

   model 3 recall: 81.25%
baseline recall: 100.00%

   model 3 precision: 13.13%
baseline precision: 8.00%


### Which model would be the best fit for this use case?

- Model 3 is the most appropriate model becuase it performed best with recall as the evaluation metric at 81%. It is optimized for recall.

Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. 

The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. 

They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. 

### Which evaluation metric would be appropriate here? 

- Precision would be appropriate because it is focused on minimizing false positives. In this case it would identify defective ducks (true positives) and minimize false positives, which in this case are costly, resulting in trips to Hawaii. 






### Which model would be the best fit for this use case?

- Model 1 is the most appropriate model as it performed best with precision as the evaluation metric at 80%.  It will minimize the false positive predictions of defects.  

In [124]:
from sklearn.metrics import classification_report

In [ ]:
sklearn.metrics.accuracy_score
sklearn.metrics.precision_score
sklearn.metrics.recall_score
sklearn.metrics.classification_report

### 4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. 

First an automated algorithm tags pictures as either a cat or a dog (Phase I). 

Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

MODEL - 
* positive - dog (target is positive)
* negative - no dog (cat) 
BASELINE - most frequent category (dog)


* TP - Dog identified, dog picture
* TN - Cat identified, cat picture
* FP - Dog identified, cat picture
* FN - Cat identified, cat picture

In [116]:
df = pd.read_csv('gives_you_paws.csv')
df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [117]:
df.describe()

actual model1 model2 model3 model4
count    5000   5000   5000   5000   5000
unique      2      2      2      2      2
top       dog    dog    cat    dog    dog
freq     3254   2937   3212   2508   4253

In [119]:
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

### Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [120]:
df['baseline'] = 'dog'

In [121]:
df.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [95]:
# Model 1

positive = 'dog'

model_accuracy = (df.model1 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model1  == positive]
model_precision = (subset.model1  == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model1  == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model 1 accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()
print(f'   model 1 recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')
print()
print(f'   model 1 precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

   model 1 accuracy: 80.74%
baseline accuracy: 65.08%

   model 1 recall: 80.33%
baseline recall: 100.00%

   model 1 precision: 89.00%
baseline precision: 65.08%


In [105]:
# Model 2

positive = 'dog'

model_accuracy = (df.model2 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model2  == positive]
model_precision = (subset.model2  == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model2  == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model 2 accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()
print(f'   model 2 recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')
print()
print(f'   model 2 precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

   model 2 accuracy: 63.04%
baseline accuracy: 65.08%

   model 2 recall: 49.08%
baseline recall: 100.00%

   model 2 precision: 89.32%
baseline precision: 65.08%


In [106]:
# model 3

positive = 'dog'

model_accuracy = (df.model3 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model3  == positive]
model_precision = (subset.model3  == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model3  == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model 3 accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()
print(f'   model 3 recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')
print()
print(f'   model 3 precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

   model 3 accuracy: 50.96%
baseline accuracy: 65.08%

   model 3 recall: 50.86%
baseline recall: 100.00%

   model 3 precision: 65.99%
baseline precision: 65.08%


In [107]:
# model 4

positive = 'dog'

model_accuracy = (df.model4 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model4  == positive]
model_precision = (subset.model4  == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model4  == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model 4 accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()
print(f'   model 4 recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')
print()
print(f'   model 4 precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

   model 4 accuracy: 74.26%
baseline accuracy: 65.08%

   model 4 recall: 95.57%
baseline recall: 100.00%

   model 4 precision: 73.12%
baseline precision: 65.08%


### - a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

model 1 accuracy: 80.74%
baseline accuracy: 65.08%
___________________________    
model 2 accuracy: 63.04%
baseline accuracy: 65.08%    
___________________________   
model 3 accuracy: 50.96%
baseline accuracy: 65.08%
___________________________ 
model 4 accuracy: 74.26%
baseline accuracy: 65.08% 

Model 1 performed best with accuracy as the evaluation metric at 81%, followed by model 4 at 74%. Both of these models performed better than the baseline using the evaluation metric of accuracy. Models 2 and 3 performed worse than the baseline, with model 2 at 63% and model 3 at 51%

### - b. Suppose you are working on a team that solely deals with dog pictures. 
- Which of these models would you recomend for Phase I? 

- First an automated algorithm tags pictures as either a cat or a dog (Phase I). 

Highest recall - Model 4 - 96% recall

- For Phase II? 
- Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Highest precision - Model 2 - 89.32% precision

### - c. Suppose you are working on a team that solely deals with cat pictures. 

Which of these models would you recomend for Phase I? 

In [ ]:
Highest recall - most positives

For Phase II?

5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

sklearn.metrics.accuracy_score
sklearn.metrics.precision_score
sklearn.metrics.recall_score
sklearn.metrics.classification_report